go/vt/vtgate/engine/routing.go

```diff
@@ -273,9 +279,8 @@ func (rp *RoutingParameters) routedTable(ctx context.Context, vcursor VCursor, b
 			// if we were able to find information about this table, let's use it
 
 			// check if the query is send to single keyspace.
-			if routedKs == nil {
-				routedKs = routedTable.Keyspace
-			}
+			once.Do(func() { routedKs = routedTable.Keyspace })
```

Comment:
@vitessio/query-serving Is this path (very) performance sensitive? Since afaik this is a more expensive way to express the same thing there and it might not be worth it if it's very performance critical?

Reply:
Good question. FWIW the implementation of once is quite small and I would expect it to be basically free (just an atomic load) following the first iteration that stores a result.

```go
func (o *Once) Do(f func()) {
	// Note: Here is an incorrect implementation of Do:
	//
	//	if atomic.CompareAndSwapUint32(&o.done, 0, 1) {
	//		f()
	//	}
	//
	// Do guarantees that when it returns, f has finished.
	// This implementation would not implement that guarantee:
	// given two simultaneous calls, the winner of the cas would
	// call f, and the second would return immediately, without
	// waiting for the first's call to f to complete.
	// This is why the slow path falls back to a mutex, and why
	// the atomic.StoreUint32 must be delayed until after f returns.

	if atomic.LoadUint32(&o.done) == 0 {
		// Outlined slow-path to allow inlining of the fast-path.
		o.doSlow(f)
	}
}

func (o *Once) doSlow(f func()) {
	o.m.Lock()
	defer o.m.Unlock()
	if o.done == 0 {
		defer atomic.StoreUint32(&o.done, 1)
		f()
	}
}
```